server/src/main/java/com/genymobile/scrcpy/util/Settings.java

```diff
@@ -16,23 +18,43 @@ private Settings() {
     public static String getValue(String table, String key) throws SettingsException {
         try (ContentProvider provider = ServiceManager.getActivityManager().createSettingsProvider()) {
             return provider.getValue(table, key);
+        } catch (Exception e) {
+            throw new SettingsException("getValue", table, key, null, e);
         }
     }
 
     public static void putValue(String table, String key, String value) throws SettingsException {
-        try (ContentProvider provider = ServiceManager.getActivityManager().createSettingsProvider()) {
-            provider.putValue(table, key, value);
-        }
+        if (Process.myUid() == 1000) {
+            // UID_SYSTEM path: fallback to 'su -c settings put ...'
+            try {
+                String cmd = "settings put " + table + " " + key + " " + value;
+                String[] command = { "su", "-c", cmd };
```

Comment:
Very vulnerable to shell injection (with root privileges). Even if `putValue()` is currently only called with safe values, this condition can be broken down easily in the future.

In the past, settings were in some conditions be set by executing a command, but with separate arguments (removed by 939c8e7f68583f444bd3f2dd4aa8b02f0ca11ddc). Hopefully this can be done also with `su`.

Reply:
I currently don't have any root device to test, what's the error with the default implementation using the settings provider with root user?